mantle/kola/harness.go

```diff
@@ -533,6 +538,58 @@ func metadataFromTestBinary(executable string) (*externalTestMeta, error) {
 	return meta, nil
 }
 
+func runExternalTest(c cluster.TestCluster, mach platform.Machine) error {
+	var previousRebootState string
+	var stdout []byte
+	var stderr []byte
+	for {
+		bootID, err := platform.GetMachineBootId(mach)
+		if err != nil {
+			return errors.Wrapf(err, "getting boot id")
+		}
+		plog.Debug("Starting kolet run-test-unit")
+		if previousRebootState != "" {
+			plog.Debugf("Setting AUTOPKGTEST_REBOOT_MARK=%s", previousRebootState)
+			c.MustSSHf(mach, "echo AUTOPKGTEST_REBOOT_MARK=%s | sudo tee /run/kola-runext-env", previousRebootState)
+			previousRebootState = ""
+		}
+		// We ignore stderr here, it will end up in the journal
+		stdout, stderr, err = mach.SSH(fmt.Sprintf("sudo ./kolet run-test-unit %s", shellquote.Join(KoletExtTestUnit)))
+		if err != nil {
+			return errors.Wrapf(err, "kolet run-test-unit failed: %s", stderr)
+		}
+		koletRes := KoletResult{}
+		if len(stdout) > 0 {
+			err = json.Unmarshal(stdout, &koletRes)
+			if err != nil {
+				return errors.Wrapf(err, "parsing kolet json %s", string(stdout))
+			}
+		}
+		if koletRes.Reboot != "" {
+			previousRebootState = koletRes.Reboot
+			plog.Debugf("Reboot request with mark='%s'", previousRebootState)
+			// Stop sshd to ensure we don't race trying to log back in during shutdown
+			_, _, err = mach.SSH(fmt.Sprintf("sudo /bin/sh -c 'systemctl stop sshd && systemctl stop %s'", KoletRebootWaitUnit))
+			if err != nil {
+				return errors.Wrapf(err, "failed to stop %s", KoletRebootWaitUnit)
+			}
+			plog.Debug("Waiting for reboot")
+			suberr := mach.WaitForReboot(120*time.Second, bootID)
+			if suberr == nil {
+				plog.Debug("Reboot complete")
+				err = nil
+				continue
+			} else {
+				return errors.Wrapf(suberr, "Waiting for reboot")
+			}
+		}
+		// Otherwise, we're done
+		if err == nil {
+			return nil
+		}
```

Comment:
Hmm, not sure if I follow the error handling here. Isn't this equivalent to e.g.:

```go
for {
  ...
  if kolaRes.Reboot == "" {
    return nil // we're done!
  }
  // code in if-statement block here; drop the `err = nil` & `continue` bits
}
```

?